src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,153 @@
+AllTypeVariant DictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  return _dictionary.at(_attribute_vector->get(chunk_offset));
+}
+
+template <typename T>
+T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
+  Assert(_attribute_vector->get(chunk_offset) != null_value_id(), "Trying to access data that points to a NULL_VALUE.");
+  return _dictionary.at(_attribute_vector->get(chunk_offset));
+}
+
+template <typename T>
+std::optional<T> DictionarySegment<T>::get_typed_value(const ChunkOffset chunk_offset) const {
+  if (_attribute_vector->get(chunk_offset) == null_value_id()) {
+    return std::nullopt;
+  }
+  return get(chunk_offset);
+}
+
+template <typename T>
+const std::vector<T>& DictionarySegment<T>::dictionary() const {
+  return _dictionary;
+}
+
+template <typename T>
+std::shared_ptr<const AbstractAttributeVector> DictionarySegment<T>::attribute_vector() const {
+  return _attribute_vector;
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::null_value_id() const {
+  return _null_value_id;
+}
+
+template <typename T>
+const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
+  return _dictionary.at(value_id);
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::value_id_for_value(const T value) const {
+  const auto index = find(_dictionary.begin(), _dictionary.end(), value);
+  Assert(index != _dictionary.end(), "Value is not contained in dictionary.");
+  return static_cast<ValueID>(std::distance(_dictionary.begin(), index));
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::lower_bound(const T value) const {
+  const auto lower_bound_position = std::lower_bound(_dictionary.begin(), _dictionary.end(), value);
+  if (lower_bound_position == _dictionary.end()) {
+    return INVALID_VALUE_ID;
+  }
+  return static_cast<ValueID>(std::distance(_dictionary.begin(), lower_bound_position));
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::lower_bound(const AllTypeVariant& value) const {
+  return lower_bound(type_cast<T>(value));
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::upper_bound(const T value) const {
+  const auto upper_bound_position = std::upper_bound(_dictionary.begin(), _dictionary.end(), value);
+  if (upper_bound_position == _dictionary.end()) {
+    return INVALID_VALUE_ID;
+  }
+  return static_cast<ValueID>(std::distance(_dictionary.begin(), upper_bound_position));
+}
+
+template <typename T>
+ValueID DictionarySegment<T>::upper_bound(const AllTypeVariant& value) const {
+  return upper_bound(type_cast<T>(value));
+}
+
+template <typename T>
+ChunkOffset DictionarySegment<T>::unique_values_count() const {
+  return _dictionary.size();
+}
+
+template <typename T>
+ChunkOffset DictionarySegment<T>::size() const {
+  return _attribute_vector->size();
+}
+
+template <typename T>
+size_t DictionarySegment<T>::estimate_memory_usage() const {
+  return _attribute_vector->width() * size() + sizeof(T) * _dictionary.size();
```

Comment:
As mentioned in class, please consider using `capacity()` instead of `size()` to get an accurate memory consumption:

```suggestion
  return _attribute_vector->width() * size() + sizeof(T) * _dictionary.capacity();
```